cloudinit/config/cc_grub_dpkg.py

```diff
@@ -43,10 +44,58 @@
 import os
 
 from cloudinit import util
+from cloudinit.util import ProcessExecutionError
 
 distros = ['ubuntu', 'debian']
 
 
+def fetch_idevs(log):
+    """
+    Fetches the /dev/disk/by-id device grub is installed to.
+    Falls back to plain disk name if no by-id entry is present.
+    """
+    disk = ""
+    devices = []
+
+    try:
+        # get the root disk where the /boot directory resides.
+        disk = util.subp(['grub-probe', '-t', 'disk', '/boot'],
+                         capture=True)[0].strip()
+    except ProcessExecutionError as e:
+        # grub-common may not be installed, especially on containers
+        # FileNotFoundError is a nested exception of ProcessExecutionError
+        if isinstance(e.reason, FileNotFoundError):
+            log.debug("grub-common is not installed, e.g. inside a container")
+        # disks from the container host are present in /proc and /sys
+        # which is where grub-probe determines where /boot is.
+        # it then checks for existence in /dev, which fails as host disks
+        # are not exposed to the container.
+        elif "failed to get canonical path" in e.stderr:
+            log.debug("Device mapping between /proc/self/mountinfo does not "
+                      "match /dev, e.g. inside a container")
+        else:
+            # something bad has happened, continue to log the error
+            raise
+    except Exception:
+        util.logexc(log, "grub-probe failed to execute for grub-dpkg")
+
+    try:
+        # check if disk exists and use udevadm to fetch symlinks
+        if os.path.exists(disk):
+            devices = util.subp(['udevadm', 'info', '-r',
+                                '--query=symlink', disk],
+                                capture=True)[0].strip().split()
+    except Exception:
+        util.logexc(log, "udevadm failed to gather devices for grub-dpkg")
+
+    # filter symlinks for /dev/disk/by-id entries
+    devices = [dev for dev in devices if 'disk/by-id' in dev]
+    # select first device if there is one, else fall back to plain name
+    idevs = sorted(devices)[0] if devices else disk
```

Comment:
I suspect we should log the output of udevadm info query, and our selection, otherwise we may not understand
why sometimes we have a by-id symlink, why it's different from run to run or when we have a plain disk path.